WIP: Implement Better Suggestions Framework

vexx32 opened the pull request:
<!-- Anything that looks like this is a comment and can't be seen after the Pull Request is created. -->

# PR Summary

Replace current niche Suggestions framework with more intuitive & simple Suggestion model that ties directly into ErrorRecord objects.

Exposing as WIP PR so that anyone wanting to give feedback can. This is in no way a concrete final implementation (yet), but hopefully it should give an idea of what we could work with here.

## Details

### Steps

- [x] Implement ~~Error~~SuggestionInfo class
- [x] Add Suggestion member to ErrorRecord and provide constructor for including Suggestions.
- [x] Replace current in-engine Suggestions with new model suggestions.
- [x] Break Suggestions relationship to ConsoleHost so that hosted use cases can leverage suggestions.
- [x] Update ErrorRecord formatting to account for when there is a suggestion.
- [x] Support multiple suggestions for same instance
- [ ] Hide everything inside experimental feature flags.

### Goals

* Keep it simple, make it useful.
* Replace current Suggestion framework as seamlessly as possible
* Use readily-available classes and methods with intuitive design for module authors who wish to utilise the framework.

## PR Context

@SteveL-MSFT mentioned that the current Suggestion model needed a rework. 🤷‍♂️ 

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: <!-- Number/link of that issue here -->
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` to your commit messages if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

vexx32 committed:
:construction: Initial framework for ErrorSuggestionInfo

vexx32 committed:
:construction: flesh out constructor somewhat

vexx32 committed:
:recycle: Use nameof() in ArgumentNullException & remove empty XML tags

vexx32 committed:
:construction: Flesh out ErrorSuggestionInfo class
:construction: Add Suggestion member as member of ErrorRecord.

vexx32 committed:
:wrench: :construction: Allow no-args suggestions

vexx32 committed:
:construction: Add SB-only opt for suggestion ctor

vexx32 committed:
:construction: Add Copy() methods to suggestions

:construction: allow >1 suggestion per ErrorRecord

:construction: Add Suggestion as a copied member
:recycle: Tidy member declarations

SteveL-MSFT commented:
Some things to think about:

* Current suggestions is tied to console host, so certainly we want to break this relationship so that hosted use cases can leverage suggestions
* Suggestions are currently hardcoded into the engine, there should be a way for modules to add suggestions
* ErrorRecord formatting needs to be updated to account for when there is a suggestion
* Support multiple suggestions (the current system supports this today, although it's not really used) for same instance

vexx32 commented:
@SteveL-MSFT Cheers, much appreciated!

Currently I'm thinking that it may make the most sense to simply require Suggestions to be supplied inside the ErrorRecord itself when creating the error record (or added post-creation by setting the property, if desired). Adding a whole bunch of logic for a suggestion-per-each-error-type model seems... unnecessarily reliant on the engine, when it could just be a simple "query the ErrorRecord to see if it's got a suggestion available" when displaying errors.

I can see an argument for wanting to keep around a suggestion item here and there in some kind of cache I suppose, but equally modules could simply keep a copy of their base Suggestion object in a module-scoped variable or static property for class-based modules. It could even be a base suggestion that you just copy around and supply new arguments to whenever needed.

I'm currently looking at how best to decouple the suggestions from conhost completely. Need to examine where CommandNotFoundException (for example) is typically generated and where the ErrorRecord for it is built. Some of the others can be implemented in their respective cmdlets themselves or perhaps in the common parameters implementations (for Transaction-related suggestions I suppose?)

Multiple suggestions wouldn't be too hard, just a matter of allowing it and working out display kinks.

Half-thinking aloud, but yeah should be pretty doable. 😄

pougetat commented:
Hey @vexx32, what would you think of creating a meta-issue to track ongoing work on this? :smiley: 
 I feel like it could be useful to track these discussions since they will hopefully result in multiple PRs and multiple people collaborating on this big work item. :smiley:
Having said that I might be overestimating the size of this work and so perhaps it isn't actually necessary. Thoughts ?

vexx32 commented:
I am hopeful to get a majority of these items done in this PR alone, if not all of them... A meta-issue isn't necessarily a _bad_ idea, either.

For the moment I'll hold off as I'm not sure it'll be needed. We can cross that bridge when it comes to it, I suppose? 😄

vexx32 commented:
@SteveL-MSFT Not really sure how to incorporate the following suggestions, partly because they don't seem to be tied to an actual _error_ of any kind. Can you provide some context on these ones? All seem to be transaction-related, but I'm not sure where they'd normally be shown? Would they just show up in _any_ error where the first command is one of the listed command names?

```powershell
NewSuggestion(
    id: 1,
    category: "Transactions",
    matchType: SuggestionMatchType.Command,
    rule: "^Start-Transaction",
    suggestion: SuggestionStrings.Suggestion_StartTransaction,
    enabled: true),
NewSuggestion(
    id: 2,
    category: "Transactions",
    matchType: SuggestionMatchType.Command,
    rule: "^Use-Transaction",
    suggestion: SuggestionStrings.Suggestion_UseTransaction,
    enabled: true),
NewSuggestion(
    id: 3,
    category: "General",
    matchType: SuggestionMatchType.Dynamic,
    rule: ScriptBlock.CreateDelayParsedScriptBlock(s_checkForCommandInCurrentDirectoryScript, isProductCode: true),
    suggestion: ScriptBlock.CreateDelayParsedScriptBlock(s_createCommandExistsInCurrentDirectoryScript, isProductCode: true),
    suggestionArgs: new object[] { CodeGeneration.EscapeSingleQuotedStringContent(SuggestionStrings.Suggestion_CommandExistsInCurrentDirectory) },
    enabled: true)
```

vexx32 committed:
:construction: Rename members more appropriately

vexx32 committed:
:construction: add suggestion to error generation
:construction: Add fuzzy match suggestion
:construction: Add "command in current dir" suggestion
:construction: Always use default suggestion
:recycle: Fix doc tags for CodeFactor

:construction: remove suggestions covered

:construction: remove unused command strings

:construction: Copy suggestions to collection

vexx32 committed:
:construction: Cache the final suggestion text

:construction: move suggestion to command

:construction: move from host to UseTransactionCommand

:recycle: rename private members for suggestions

:recycle: Tidy suggestions class

:recycle: Remove accounted-for suggestion

:recycle: Remove from HostUtilities

vexx32 committed:
:construction: Add implicit string operator

vexx32 committed:
:construction: Add suggestion to error display

:construction: Account for errors in suggestions
:construction: Avoid throwing errors during suggestion display

iSazonov commented:
My cent. We could consider decoupling suggestions in separate module/dll (internal at first).

vexx32 commented:
Hmm. We *could* do that, but while I can easily decouple the suggestion *items*, I don't really see an effective path to decouple the suggestion framework itself.

Currently, I'm attaching it to ErrorRecord in order to have it easily accessible both to the engine and module authors.

I'll revisit this periodically as I work to see if I can think of a good way to accomplish this. :smile:

iSazonov commented:
This could be in Goals.

I'd expect that suggestion system reacts to different events and it is not tied exclusively to ErrorRecord.

vexx32 commented:
I can definitely agree that could potentially be useful, but I think ultimately that would most likely create confusion as to what's being output / what's being *suggested*.

The alternative, I suppose, would be to make it properly a kind of InformationRecord that is shown or hidden based on a preference variable of some kind, determined separately to other information records.

This would mean whenever you want to add a suggestion, you need to call `WriteInformation()` _as well as_ `WriteError()` in a cmdlet. In script modules we would need to either utilise Write-Information _or_ we would need to create a new Write-Suggestion cmdlet.

What do you think of this approach?

iSazonov commented:
I can not think a scenario in which we could use InformationRecord but it could be.
This is a special case of my proposal - the system must be able to respond to any events: ErrorRecords, heavy resource consumption (memory, CPU), large remote (de)serialization...

vexx32 commented:
I agree that potentially those could be useful suggestions to include.

However, I do not know of any place in the engine that currently monitors such events. I think it would be better to implement the generic case and leave it open-ended for both module authors and the engine to add cases as they need.

For this case, I think the generic ErrorRecord or a type inheriting InformationRecord would be best. Then specific suggestions can be added wherever is appropriate.

vexx32 committed:
:construction: Order & rename members properly

vexx32 committed:
:construction: Remove 90% of the base suggestions

vexx32 committed:
:construction: rename suggestion class & cleanup

SteveL-MSFT commented:
@vexx32 Transactions aren't supported in PSCore6, so you should just remove those existing suggestions.  If the suggestion (pardon the pun) is to output the suggestion to the information stream, I'm not sure if that is useful in this case.  Suggestions are meant for interactive use.  Perhaps we need a separate member for SuggestedFix which would be useful in other cases (like PSReadLine could read that and inline the fix).

vexx32 commented:
@SteveL-MSFT I was not aware of that. Are you sure? The cmdlets `Start/Stop/Using-Transaction` all seem to be present at least in the codebase still.

Agreed we could potentially open up such an avenue for PSReadLine's use, not sure if it makes sense necessarily to do in this same PR?

One problem I'm having at the moment is that although I've added the additional code to the formatview for ErrorRecord, I don't see it getting written to the console when a generated ErrorRecord is displayed. I've verified that the ErrorRecord contains the expected suggestion from the CommandNotFoundException, but it still isn't showing up in the console.

Is there something unusual about ErrorRecord's formatview I'm missing here? 😄

SteveL-MSFT commented:
@vexx32 The code is still there because it's inherited from Windows PowerShell, but if you look in the csproj, those source files are excluded from the build.

I think we can split the entirety of the suggestions framework work to multiple PRs, so exposing `SuggestedFix` can certainly be separate from this. 

You can use `Format-List * -Force` to force ErrorRecord not to use it's default formatter.

vexx32 commented:
Gotcha.

Yeah, I'm trying to work the suggestion into the default formatter, though. I can see the member present with the correct data, just the changes I made to the formatter for ErrorRecord don't seem to be taking effect, and I don't see anything else that modifies its formatter in the repo. :/

RDIL commented on src/System.Management.Automation/engine/ErrorPackage.cs:
```diff
@@ -184,6 +185,117 @@ public enum ErrorCategory
+                    "Transaction",
+                    new SuggestionInfo(
+                        ScriptBlock.CreateDelayParsedScriptBlock(
+                            SuggestionStrings.Suggestion_StartTransaction, isProductCode: true))
```
```suggestion
                            SuggestionStrings.Suggestion_StartTransaction, isProductCode: true
                        )
                    )
```

vexx32 replied:
Personally, _I_ certainly prefer this format, but it's not a style generally used in this repo. 😄

vexx32 committed:
:recycle: Remove unneeded suggestions

vexx32 committed:
:wrench: Resolve suggestion text during ctor
:memo: This is needed to allow enumeration of Suggestion collections

vexx32 committed:
:recycle: Rework suggestion display format

vexx32 commented:
@SteveL-MSFT figured out what I was breaking there, it wasn't liking how I set up the getter to modify the object property in-place when it was iterating over a collection.

Current format looks like this: (first is a suggestion from CommandNotFoundException, second is a custom one with two suggestions as a quick test)

```
PS C:\Users\Joel\AppData\Local\Temp> throw $e
hi
At line:1 char:1
+ throw $e
+ ~~~~~~~~
+ CategoryInfo          : ConnectionError: (:) [], Exception
+ FullyQualifiedErrorId : testID
 
Suggestions:
  Hello, this is a test suggestion. Please adjust your headset.
  Thank you, test complete. Please remove the diskette.
```

```
PS C:\Users\Joel\AppData\Local\Temp> command2
command2 : The term 'command2' is not recognized as the name of a cmdlet, function, script file, or operable program.
Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
At line:1 char:1
+ command2
+ ~~~~~~~~
+ CategoryInfo          : ObjectNotFound: (command2:String) [], CommandNotFoundException
+ FullyQualifiedErrorId : CommandNotFoundException
 
Suggestions:
  The command command2 was not found, but does exist in the current location. PowerShell does not load commands from the current location by default. If you trust this command, instead type: ".\command2". See "get-help about_Command_Precedence" for more details.
```

Screenshots, since the red doesn't carry through that well in Markdown:

![image](https://user-images.githubusercontent.com/32407840/54952471-7610c580-4f1c-11e9-98d7-03d3430d0447.png)

![image](https://user-images.githubusercontent.com/32407840/54952485-7f019700-4f1c-11e9-81de-9f4bce64373f.png)

vexx32 committed:
:recycle: Remove mess from rebase

SteveL-MSFT commented on src/System.Management.Automation/FormatAndOutput/DefaultFormatters/PowerShellCore_format_ps1xml.cs:
```diff
@@ -777,62 +777,67 @@ private static IEnumerable<FormatViewDefinition> ViewsOf_System_Management_Autom
+                                        if ($_.Suggestions.Count -gt 0) {
+                                            $Prefix = "" "" * 2
+                                            $NewLine
+                                            ""Suggestions:""
```
Keeping the current rendering format, the suggestion should be a different color

vexx32 replied:
I don't entirely disagree, but I don't think the suggestion should look like the output. I'm more okay with it looking like it's part of the error than I am with it looking like it's stray output. That might encourage new users to simply drop random text strings to output with their own suggestions rather than investigating how to do it more appropriately, creating an issue of needlessly cluttering the output stream with jagged collections of various object types.

SteveL-MSFT replied:
I don't think it has to be white like it is today which I agree looks like regular output.  Perhaps a nice green color to make it positive :)

vexx32 replied:
Would that be best done with terminal escape sequences?

SteveL-MSFT replied:
In my opinion, the terminal world (including Windows now) has accepted escape sequences as the standard so I would say yes.

vexx32 replied:
Alright, guess I'll go hunting for a suitable green color then!

vexx32 replied:
@SteveL-MSFT so I figured out how to do that, and it's nice.

However, given that console colors can be customized... this needs to be configurable, doesn't it? If someone (I mean, I wouldn't ever want to see it, but it could happen) chose a green background for their console, they'd never see suggestions again. Similarly confusing if they elected to change their default foreground color to some shade of green for "that hacker look".

Should there be an environment variable or something for this?

SteveL-MSFT replied:
@vexx32 I think customization of colors should be discussed here: https://github.com/PowerShell/PowerShell/issues/3611 outside of this PR.  Once we have that solved, we will need to update any code that has raw escape sequences.  In the future, I'd also like to see markdown rendering of strings, so you could just use markdown to add emphasis.  However, none of this exists today.

vexx32 committed:
:sparkles: add color to suggestions

rkeithhill commented:
Rather than picking **a** color, you could try to invert the suggestions e.g.:

![image](https://user-images.githubusercontent.com/5177512/56709130-e3ab3f80-66dc-11e9-8384-9a4da173b598.png)

![image](https://user-images.githubusercontent.com/5177512/56709156-fa519680-66dc-11e9-9167-2c5b96045dc5.png)

vexx32 commented:
@rkeithhill that _would_ be a decent idea. However, the typical format rendering of ErrorRecord happens as it comes from the error stream, meaning the mostly likely colors that'll get rendered in black-on-red by default.

Not especially great for the intention here, unfortunately. 😕

rkeithhill commented:
Did that change?  On 6.2 the suggestion shows in a normal foreground color.  And I can catch the command not found exception but the suggestion text still appears.  Seems like it is writing directly to the console.  In fact, you can't seem to redirect it.  ??

vexx32 commented:
Yeppers, it is, which is one of the reasons I'd rather it be where I have it at the moment; it's very hard to manage at all.

In the current release, it writes directly to console. In fact, it's the conhost that handles them completely I think. I moved the suggestions into the ErrorRecord class as a subclass (SuggestionInfo) and then changed the ErrorRecord formatter to incorporate them. As a result... red by default.

That could all be done differently, potentially, but it seems to be the most sensible way to go and the most accessible to module developers to add their own as needed / desired.

rkeithhill commented:
I like being able to manage that display (ideally it would only happen "interactively").  But this is troubling:

> As a result... red by default.

There is so much red text I'm pretty sure a lot of folks will just miss the suggestion text.  After read a few lines of informative text and then a few lines of goobledy gook, I think most folks stop reading further.

vexx32 commented:
@rkeithhill I would generally prefer that, but it would mean somehow writing to the information stream from the error stream / formatter, or somehow detecting that there's a suggestion when an error occurs and writing to the information stream directly. 

That latter option I feel is kinda undesirable because it means there's no discoverability there. The information stream is largely just text, and I can't really think of a good way to write to information when an error occurs, short of the original implementation which is entirely reliant on conhost to make it work, meaning suggestions don't exist in non-standard host setups.

It makes the most sense, I feel, to couple it with ErrorRecord. That it also makes sense to change the colour for visibility and clarity of intent complicates things slightly, but not to the point of making it a good idea to move it elsewhere entirely, I think.

vexx32 commented:
I need to get back to this... @SteveL-MSFT I think all I have remaining here is to package it inside experimental feature flags, unless you have further comments on the implementation thus far. 🙂

vexx32 commented:
Hah, nope. Gotta add some. After I fix whatever's still broken. I'll see if I can do a rebase and then fix up whatever broke soon.

Do you guys want me to wrap this up as an experimental feature, or plan for it to be a PS7 full feature?

vexx32 commented:
@PoshChan please remind me in 2 days.

PoshChan commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs:
```diff
@@ -2357,7 +2357,6 @@ internal void Run(bool inputLoopIsNested)
                 Dbg.Assert(ui != null, "Host.UI should return an instance.");
```
@vexx32, your last commit had 2 failures in `PowerShell-CI-linux`
PowerShellGet - Module tests (Admin).Error occurred in Describe block
```powershell
At line:2 char:57
+             [System.Diagnostics.DebuggerHiddenAttribute()]
+                                                         ~
Unexpected token ')' in expression or statement.
At line:8 char:34
```
PowerShellGet - Script tests (Admin).Should install a script correctly to the required location with AllUsers scope
```powershell
At line:2 char:57
+             [System.Diagnostics.DebuggerHiddenAttribute()]
+                                                         ~
Unexpected token ')' in expression or statement.
at <ScriptBlock>, : line 8
8:             $foundSuggestion = @(Get-Command ($ExecutionContext.SessionState.Path.Combine(".", $escapedCommand)) -ErrorAction Ignore).Count -gt 0

```

PoshChan replied:
@vexx32, your last commit had 2 failures in `PowerShell-CI-linux`
PowerShellGet - Module tests (Admin).Error occurred in Describe block
```powershell
At line:2 char:57
+             [System.Diagnostics.DebuggerHiddenAttribute()]
+                                                         ~
Unexpected token ')' in expression or statement.
At line:8 char:34
```
PowerShellGet - Script tests (Admin).Should install a script correctly to the required location with AllUsers scope
```powershell
At line:2 char:57
+             [System.Diagnostics.DebuggerHiddenAttribute()]
+                                                         ~
Unexpected token ')' in expression or statement.
at <ScriptBlock>, : line 8
8:             $foundSuggestion = @(Get-Command ($ExecutionContext.SessionState.Path.Combine(".", $escapedCommand)) -ErrorAction Ignore).Count -gt 0

```

PoshChan replied:
@vexx32, test results were not published for PowerShell-CI-macos at vstfs:///Build/Build/26670

PoshChan replied:
@vexx32, test results were not published for PowerShell-CI-static-analysis at vstfs:///Build/Build/26669

PoshChan replied:
@vexx32, test results were not published for PowerShell-CI-windows at vstfs:///Build/Build/26672

PoshChan replied:
@vexx32, your last commit had 2 failures in `PowerShell-CI-linux`
PowerShellGet - Module tests (Admin).Error occurred in Describe block
```powershell
At line:2 char:57
+             [System.Diagnostics.DebuggerHiddenAttribute()]
+                                                         ~
Unexpected token ')' in expression or statement.
At line:8 char:34
```
PowerShellGet - Script tests (Admin).Should install a script correctly to the required location with AllUsers scope
```powershell
At line:2 char:57
+             [System.Diagnostics.DebuggerHiddenAttribute()]
+                                                         ~
Unexpected token ')' in expression or statement.
at <ScriptBlock>, : line 8
8:             $foundSuggestion = @(Get-Command ($ExecutionContext.SessionState.Path.Combine(".", $escapedCommand)) -ErrorAction Ignore).Count -gt 0

```

PoshChan replied:
@vexx32, your last commit had 2 failures in `PowerShell-CI-macos`
Native streams behavior with PowerShell.Error stream.preserves error stream as is with Out-String
```powershell
Expected strings to be the same, but they were different.
Expected length: 33
Actual length:   51
Strings differ at index 4.
Expected: 'foo\n\nbar\n\nbazmiddlefoo\n\nbar\n\nbaz\n'
But was:  'foo\n \n\n \nb...'
at <ScriptBlock>, /Users/vsts/agent/2.153.2/work/1/s/test/powershell/Language/Scripting/NativeExecution/NativeStreams.Tests.ps1: line 52
52:             ($out | Out-String).Replace("`r", '') | Should -BeExactly "foo`n`nbar`n`nbazmiddlefoo`n`nbar`n`nbaz`n"
```
Native streams behavior with PowerShell.Error stream.does not get truncated or split when redirected
```powershell
Expected 1, but got 2.
at <ScriptBlock>, /Users/vsts/agent/2.153.2/work/1/s/test/powershell/Language/Scripting/NativeExecution/NativeStreams.Tests.ps1: line 62
62:             $e.Count | Should -Be 1
```

PoshChan replied:
@vexx32, your last commit had 4 failures in `PowerShell-CI-windows`
Tests module table restrictions.Verifies that Get-Command does not expose private module function under system lock down
```powershell
Cannot invoke method. Method invocation is supported only on core types in this language mode.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Security\ConstrainedLanguageModules.Tests.ps1: line 559
559:                 $GetCommandPrivateFnCmdInfo = Get-Command -Name "PrivateFn1" 2> $null

```
Tests module table restrictions.Verifies that Get-Command does not expose private function after explicitly importing nested module file under system lock down
```powershell
Cannot invoke method. Method invocation is supported only on core types in this language mode.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Security\ConstrainedLanguageModules.Tests.ps1: line 593
593:                 $ReImportPrivateFnCmdInfo = Get-Command -Name "PrivateFn1" 2> $null

```
Native streams behavior with PowerShell.Error stream.preserves error stream as is with Out-String
```powershell
Expected strings to be the same, but they were different.
Expected length: 33
Actual length:   51
Strings differ at index 4.
Expected: 'foo\n\nbar\n\nbazmiddlefoo\n\nbar\n\nbaz\n'
But was:  'foo\n \n\n \nb...'
at <ScriptBlock>, D:\a\1\s\test\powershell\Language\Scripting\NativeExecution\NativeStreams.Tests.ps1: line 52
52:             ($out | Out-String).Replace("`r", '') | Should -BeExactly "foo`n`nbar`n`nbazmiddlefoo`n`nbar`n`nbaz`n"
```
Native streams behavior with PowerShell.Error stream.does not get truncated or split when redirected
```powershell
Expected 1, but got 2.
at <ScriptBlock>, D:\a\1\s\test\powershell\Language\Scripting\NativeExecution\NativeStreams.Tests.ps1: line 62
62:             $e.Count | Should -Be 1
```

PoshChan replied:
@vexx32, your last commit had 2 failures in `PowerShell-CI-linux`
Native streams behavior with PowerShell.Error stream.preserves error stream as is with Out-String
```powershell
Expected strings to be the same, but they were different.
Expected length: 33
Actual length:   51
Strings differ at index 4.
Expected: 'foo\n\nbar\n\nbazmiddlefoo\n\nbar\n\nbaz\n'
But was:  'foo\n \n\n \nb...'
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Language/Scripting/NativeExecution/NativeStreams.Tests.ps1: line 52
52:             ($out | Out-String).Replace("`r", '') | Should -BeExactly "foo`n`nbar`n`nbazmiddlefoo`n`nbar`n`nbaz`n"
```
Native streams behavior with PowerShell.Error stream.does not get truncated or split when redirected
```powershell
Expected 1, but got 2.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Language/Scripting/NativeExecution/NativeStreams.Tests.ps1: line 62
62:             $e.Count | Should -Be 1
```

vexx32 commented:
@PoshChan please get failures.

PoshChan commented:
@vexx32, this is the reminder you requested 2 days. ago

vexx32 commented:
@daxian-dbw @SteveL-MSFT I've hit a bit of a wall here, it seems. Somehow the changes to the CommandNotFoundException is causing an infinite loop. Didn't think I'd made any significant changes to the logic of it, and debugging it is leading me... essentially nowhere (just round and round the loop, very tricky to track what the heck is going on that's causing it).

If y'all have any ideas as to what the heck is happening, I'd appreciate it. 💖 

Doubly weird is that it's not the same on Linux? It seems to register a parsing error in a DebuggerHidden attribute, but I really have no idea why it would do that... best guess is a strange error where it's exceeding the call stack limitations and is just totally falling to pieces.

vexx32 committed:
:wrench: Add short-circuit to command discovery
Without this change, PS just recurses whenever a command isn't found.

vexx32 committed:
:wrench: Fix suggestion header
Header is prevented from showing if there are no visible suggestions

vexx32 commented:
@TravisEz13 where should I put tests for this?

Commit:
test

vexx32 committed:
test

SteveL-MSFT commented:
@vexx32 can you resolve the merge conflict?  I'd like to spend some time on this and see if we can get it into 7.1 preview.1

vexx32 committed:
Merge branch 'master' into BetterSuggestions

vexx32 commented:
@SteveL-MSFT I think that ought to do it. 🙂

PoshChan commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs:
```diff
@@ -2419,7 +2419,6 @@ internal void Run(bool inputLoopIsNested)
                 Dbg.Assert(ui != null, "Host.UI should return an instance.");
```
@vexx32, your last commit had 41 failures in `PowerShell-CI-linux`
(These are 5 of the failures)

Online help tests for PowerShell Cmdlets.Error occurred in Describe block
```powershell
Cannot validate argument on parameter 'Name'. The argument is null or empty. Provide an argument that is not null or empty, and then try the command again.
At line:16 char:34
```
PowerShellGet - Module tests (Admin).Error occurred in Describe block
```powershell
The variable '$lastError' cannot be retrieved because it has not been set.
At line:16 char:34
```
PowerShellGet - Script tests (Admin).Should install a script correctly to the required location with AllUsers scope
```powershell
The variable '$lastError' cannot be retrieved because it has not been set.
at <ScriptBlock>, : line 16
16:             $foundSuggestion = @(Get-Command ($ExecutionContext.SessionState.Path.Combine(".", $escapedCommand)) -ErrorAction Ignore).Count -gt 0

```
ConsoleHost unit tests.ShellInterop.Verify Validate Dollar Error Populated should throw exception
```powershell
Expected an exception, with FullyQualifiedErrorId 'CommandNotFoundException' to be thrown, but the FullyQualifiedErrorId was 'ParameterArgumentValidationError,Microsoft.PowerShell.Commands.GetCommandCommand'. from /home/vsts/work/1/s/test/powershell/Host/ConsoleHost.Tests.ps1:108 char:22
    + ?        $a | & $powershell -noprofile -command { wgwg-wrwrhqwrhrh35h3h ?
    +               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Host/ConsoleHost.Tests.ps1: line 109
109:             } | Should -Throw -ErrorId "CommandNotFoundException" -PassThru
```
ConsoleHost unit tests.Pipe to/from powershell.errors are in text if error is redirected, encoded command, non-interactive, and outputformat specified
```powershell
Expected a value, but got $null or empty.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Host/ConsoleHost.Tests.ps1: line 416
416:             $out | Should -Not -BeNullOrEmpty
```

PoshChan replied:
@vexx32, your last commit had 40 failures in `PowerShell-CI-macos`
(These are 5 of the failures)

Online help tests for PowerShell Cmdlets.Error occurred in Describe block
```powershell
Cannot validate argument on parameter 'Name'. The argument is null or empty. Provide an argument that is not null or empty, and then try the command again.
At line:16 char:34
```
PowerShellGet - Module tests (Admin).Error occurred in Describe block
```powershell
The variable '$lastError' cannot be retrieved because it has not been set.
At line:16 char:34
```
PowerShellGet - Script tests (Admin).Should install a script correctly to the required location with AllUsers scope
```powershell
The variable '$lastError' cannot be retrieved because it has not been set.
at <ScriptBlock>, : line 16
16:             $foundSuggestion = @(Get-Command ($ExecutionContext.SessionState.Path.Combine(".", $escapedCommand)) -ErrorAction Ignore).Count -gt 0

```
ConsoleHost unit tests.ShellInterop.Verify Validate Dollar Error Populated should throw exception
```powershell
Expected an exception, with FullyQualifiedErrorId 'CommandNotFoundException' to be thrown, but the FullyQualifiedErrorId was 'ParameterArgumentValidationError,Microsoft.PowerShell.Commands.GetCommandCommand'. from /Users/runner/runners/2.164.7/work/1/s/test/powershell/Host/ConsoleHost.Tests.ps1:108 char:22
    + ?        $a | & $powershell -noprofile -command { wgwg-wrwrhqwrhrh35h3h ?
    +               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
at <ScriptBlock>, /Users/runner/runners/2.164.7/work/1/s/test/powershell/Host/ConsoleHost.Tests.ps1: line 109
109:             } | Should -Throw -ErrorId "CommandNotFoundException" -PassThru
```
ConsoleHost unit tests.Pipe to/from powershell.errors are in text if error is redirected, encoded command, non-interactive, and outputformat specified
```powershell
Expected a value, but got $null or empty.
at <ScriptBlock>, /Users/runner/runners/2.164.7/work/1/s/test/powershell/Host/ConsoleHost.Tests.ps1: line 416
416:             $out | Should -Not -BeNullOrEmpty
```

PoshChan replied:
@vexx32, your last commit had 89 failures in `PowerShell-CI-windows`
(These are 5 of the failures)

Online help tests for PowerShell Cmdlets.Error occurred in Describe block
```powershell
Cannot validate argument on parameter 'Name'. The argument is null or empty. Provide an argument that is not null or empty, and then try the command again.
At line:16 char:34
```
ConsoleHost unit tests.ShellInterop.Verify Validate Dollar Error Populated should throw exception
```powershell
Expected an exception, with FullyQualifiedErrorId 'CommandNotFoundException' to be thrown, but the FullyQualifiedErrorId was 'ParameterArgumentValidationError,Microsoft.PowerShell.Commands.GetCommandCommand'. from D:\a\1\s\test\powershell\Host\ConsoleHost.Tests.ps1:108 char:22
    + ?        $a | & $powershell -noprofile -command { wgwg-wrwrhqwrhrh35h3h ?
    +               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
at <ScriptBlock>, D:\a\1\s\test\powershell\Host\ConsoleHost.Tests.ps1: line 109
109:             } | Should -Throw -ErrorId "CommandNotFoundException" -PassThru
```
ConsoleHost unit tests.-Login pwsh switch.Error occurred in Context block
```powershell
Cannot validate argument on parameter 'Name'. The argument is null or empty. Provide an argument that is not null or empty, and then try the command again.
At line:16 char:34
```
ConsoleHost unit tests.Pipe to/from powershell.errors are in text if error is redirected, encoded command, non-interactive, and outputformat specified
```powershell
Expected a value, but got $null or empty.
at <ScriptBlock>, D:\a\1\s\test\powershell\Host\ConsoleHost.Tests.ps1: line 416
416:             $out | Should -Not -BeNullOrEmpty
```
Additional tests for Import-Module with WinCompat.Tests that ErrorAction/WarningAction have effect when Import-Module with WinCompat is used.Verify that Error is generated with default ErrorAction
```powershell
Expected 'divide by zero' to be found in file 'C:\Users\VssAdministrator\AppData\Local\Temp\6655aaa5-fed6-4160-ba8f-f1b718dc162d\4d6749bd-baa4-4f4e-bef9-b1f477bd0748', but it was not found.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Core\CompatiblePSEditions.Module.Tests.ps1: line 428
428:             $LogPath | Should -FileContentMatch 'divide by zero'
```

iSazonov commented:
How sad it is to see steel a longtime PR. :-(

jazzdelightsme commented:
I'd still like to see this happen... what is this about "shell automation" moving this from "to do" to "done"? Wouldn't it be better to leave it as "to do", or have a "stale" category instead of "done"?

vexx32 commented:
The code here is mostly good to go. The only real difficult bit is incorporating the suggestions into the ErrorRecord formatter again; the changes required for ConciseView mean that that portion of the code has to be put in from scratch in the newer code.

I know @SteveL-MSFT expressed interest in possibly picking this up, but I'm guessing the PS folks have been too busy / focused on other things thus far.

If you're interested in picking it up yourself I can point you at what would still need fixing if you'd like to take a stab at it. 🙂 

Mind you, wouldn't be bad to leave the `Stale` label so folks can sift through old PRs that never got finished and get them sorted later down the line potentially. @TravisEz13, is that something worth looking at? 🙂

iSazonov commented:
> The only real difficult bit is incorporating the suggestions into the ErrorRecord formatter again; the changes required for ConciseView mean that that portion of the code has to be put in from scratch in the newer code.

The PR is WIP only due to this? If so we could split the work I hope.

vexx32 commented:
It might be possible... However, part of the wider code changes is removing the existing suggestions (which were implemented in the default host and are effectively inaccessible to everything else), so if we merge that without having the formatter updated we will have a build where suggestions appear to be missing (unless you inspect the ErrorRecords manually).

iSazonov commented:
I see you have already changed ConciseView - is it not finished jet?

vexx32 commented:
I don't think I had time to properly finish it, no. 😕 

Could be I just forgot that I did in, in which case feel free to poach the code and attempt a rebase. I remember that being a bit hairy on the formatter file though 😓

iSazonov commented:
It's a shame to lose it...

SteveL-MSFT commented:
Closing this in favor of the new FeedbackProvider
